File: BEmu_cpp/src/ReferenceDataRequest/ReferenceElementString.cpp
//------------------------------------------------------------------------------
// <copyright project="BEmu_cpp" file="src/ReferenceDataRequest/ReferenceElementString.cpp" company="Jordan Robinson">
//     Copyright (c) 2013 Jordan Robinson. All rights reserved.
//
//     The use of this software is governed by the Microsoft Public License
//     which is included with this distribution.
// </copyright>
//------------------------------------------------------------------------------

#include "ReferenceDataRequest/ReferenceElementString.h"
#include "BloombergTypes/Name.h"
#include "Types/IndentType.h"

namespace BEmu
{
	namespace ReferenceDataRequest
	{
		ReferenceElementString::ReferenceElementString(const std::string& name, const std::string& value) : CanConvertToStringType(value)
		{
			this->_name = name;
			this->_value = value;
		}

		Name ReferenceElementString::name() const { return Name(this->_name.c_str()); }
		size_t ReferenceElementString::numValues() const { return 1; }
		size_t ReferenceElementString::numElements() const { return 0; }

		bool ReferenceElementString::isNull() const { return false; }
		bool ReferenceElementString::isArray() const { return false; }
		bool ReferenceElementString::isComplexType() const { return false; }

		void ReferenceElementString::setName(const std::string& name)
		{
			this->_name = name;
		}

		void ReferenceElementString::setValue(const std::string& value)
		{
			this->_value = value;
		}

		const char * ReferenceElementString::getValueAsString(int index) const
		{
			if(index == 0)
				return this->ValueAsString().c_str();
			else
				throw elementPtrEx;
		}

		std::ostream& ReferenceElementString::print(std::ostream& stream, int level, int spacesPerLevel) const
		{
			std::string tabs(IndentType::Indent(level, spacesPerLevel));
			stream << tabs << this->_name << " = \"" << this->_value << "\"" << std::endl;
			return stream;
		}

	}
}

File: BEmu/ReferenceDataRequest/ReferenceElementString.cs
//------------------------------------------------------------------------------
// <copyright project="BEmu_csh" file="ReferenceDataRequest/ReferenceElementString.cs" company="Jordan Robinson">
//     Copyright (c) 2013 Jordan Robinson. All rights reserved.
//
//     The use of this software is governed by the Microsoft Public License
//     which is included with this distribution.
// </copyright>
//------------------------------------------------------------------------------

namespace Bloomberglp.Blpapi.ReferenceDataRequest
{
    using System;
    using System.Collections.Generic;
    using System.Linq;
    using System.Text;

    internal class ReferenceElementString : Element
    {
        private readonly string _name, _value;

        internal ReferenceElementString(string name, string value)
        {
            this._name = name;
            this._value = value;
        }

        public override Name Name { get { return new Name(this._name); } }
        public override int NumValues { get { return 1; } }
        public override int NumElements { get { return 0; } }
        public override bool IsArray { get { return false; } }
        public override bool IsComplexType { get { return false; } }
        public override Schema.Datatype Datatype { get { return Schema.Datatype.STRING; } }

        internal override StringBuilder PrettyPrint(int tabIndent, bool surroundValueWithQuotes = false)
        {
            string valueToUse = surroundValueWithQuotes ? string.Format("\"{0}\"", this._value) : this._value;
            return base.PrettyPrintHelper(tabIndent, valueToUse);
        }

        public override object GetValue()
        {
            return this._value;
        }

        public override string GetValueAsString()
        {
            return this._value;
        }

        public override string GetValueAsString(int index)
        {
            if (index == 0)
                return this._value;
            else
                return base.GetValueAsString(index);
        }
        
    }
}
